src/components/NavBar/ChainSwitcher.tsx

```diff
@@ -0,0 +1,178 @@
+import { useLocation, useNavigate } from 'react-router-dom'
+import { addPopup } from 'state/application/reducer'
+import { updateConnectionError } from 'state/connection/reducer'
+import { useAppDispatch } from 'state/hooks'
+import { replaceURLParam } from 'utils/routes'
+import { isChainAllowed, switchChain } from 'utils/switchChain'
+
+import * as styles from './ChainSwitcher.css'
+import { NavDropdown } from './NavDropdown'
+
+const ChainRow = ({
+  targetChain,
+  onSelectChain,
+}: {
+  targetChain: SupportedChainId
+  onSelectChain: (targetChain: number) => void
+}) => {
+  const { provider, chainId } = useWeb3React()
+  if (!provider || !chainId) {
+    return null
+  }
+  const active = chainId === targetChain
+  const { label, logoUrl } = getChainInfo(targetChain)
+
+  return (
+    <Row
+      as="button"
+      background={active ? 'lightGrayContainer' : 'none'}
+      className={`${styles.ChainSwitcherRow} ${subhead}`}
+      onClick={() => onSelectChain(targetChain)}
+    >
+      <ChainDetails>
+        <img src={logoUrl} alt={label} className={styles.Icon} />
+        {label}
+      </ChainDetails>
+      {active && <CheckMarkIcon width={20} height={20} />}
+    </Row>
+  )
+}
+
+const ChainDetails = ({ children }: { children: ReactNode }) => <Row>{children}</Row>
+
+const getParsedChainId = (parsedQs?: ParsedQs) => {
+  const chain = parsedQs?.chain
+  if (!chain || typeof chain !== 'string') return { urlChain: undefined, urlChainId: undefined }
+
+  return { urlChain: chain.toLowerCase(), urlChainId: getChainIdFromName(chain) }
+}
+
+const getChainIdFromName = (name: string) => {
+  const entry = Object.entries(CHAIN_IDS_TO_NAMES).find(([_, n]) => n === name)
+  const chainId = entry?.[0]
+  return chainId ? parseInt(chainId) : undefined
+}
+
+const getChainNameFromId = (id: string | number) => {
+  // casting here may not be right but fine to return undefined if it's not a supported chain ID
+  return CHAIN_IDS_TO_NAMES[id as SupportedChainId] || ''
+}
+
+const NETWORK_SELECTOR_CHAINS = [
+  SupportedChainId.MAINNET,
+  SupportedChainId.POLYGON,
+  SupportedChainId.OPTIMISM,
+  SupportedChainId.ARBITRUM_ONE,
+]
+
+interface ChainSwitcherProps {
+  isMobile?: boolean
+}
+
+export const ChainSwitcher = ({ isMobile }: ChainSwitcherProps) => {
+  const dispatch = useAppDispatch()
+  const { chainId, provider, connector } = useWeb3React()
+  const previousChainId = usePrevious(chainId)
+  const [isOpen, toggleOpen] = useReducer((s) => !s, false)
+  const parsedQs = useParsedQueryString()
+  const { urlChain, urlChainId } = getParsedChainId(parsedQs)
+  const previousUrlChainId = usePrevious(urlChainId)
+
+  const ref = useRef<HTMLDivElement>(null)
+  useOnClickOutside(ref, isOpen ? toggleOpen : undefined)
+
+  const navigate = useNavigate()
+  const location = useLocation()
+  const info = chainId ? getChainInfo(chainId) : undefined
+
+  const onSelectChain = useCallback(
```

Comment:
this logic is shared with the existing NetworkSelector. can we move it into a separate hook so it can be re-used? https://github.com/Uniswap/widgets/pull/70 example in here